### PR TITLE
HIP-721: 06026 - add software version to events

### LINKED ISSUE #6026: HIP 721: Add software version to events
Events must track the software version used to create them for application handling purposes. If transactions are submitted to the platform close to an upgrade and do not reach consensus prior, they may reach consensus after (once pre-consensus event stream is complete). The upgrade process currently takes several minutes, during which the transaction submitter may see their transaction timeout.

The software version must be serialized when the event is written to disk and when gossiped. The former will impact consumers of the consensus event stream.

Adding the software version to the event allows the application to identify these transactions and send appropriate cancellation messages or take any other necessary action. Therefore, the `ConsensusEvent` interface must have a new method for providing the software version.

Fulfills [HIP 721](https://github.com/hashgraph/hedera-improvement-proposal/blob/main/HIP/hip-721.md)

### PR DESCRIPTION
**Pull request by edward-swirldslabs**
**Completes [HIP-721](https://github.com/hashgraph/hedera-improvement-proposal/blob/main/HIP/hip-721.md)**

**Description**:
- Adds `softwareVersion` field to BaseEventHashData.java with serialization and deserialization
- Adds nullable `getSoftwareVersion` method to Event, EventImpl, and BaseEventHashData
- All event construction requires non-null SoftwareVersion.

**Related issue(s)**:

Fixes #6026 

**Notes for reviewer**:
The return value for the `getSoftwareVersion` method is nullable for backwards compatibility, but all new event construction requires software version to be `@NonNull`

### CODE REVIEW

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/internal/EventImpl.java
Comment (hendrikebbers) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/internal/EventImpl.java:1006]: I ask myself if we should add @Nullable to an implementation if the interface already defines it. Any thoughts on that?

File: platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/main/java/com/swirlds/platform/test/simulated/SimulatedEventCreationNode.java
Comment (hendrikebbers) [inline platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/main/java/com/swirlds/platform/test/simulated/SimulatedEventCreationNode.java:87]: Really like how you update all he constructors

**Review by hendrikebbers — DISMISSED** [review]

**Review by poulok — CHANGES_REQUESTED** [review]
We can get this code ready, but let's not merge until the HIP is approved

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java
Comment (cody-littley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:229]: Perhaps now is a good time to remove this `int` from the serialization protocol, since we are ignoring its value. We can piggy-back off of the new version you added to do the necessary data migration.

We would have to add a small section to the HIP, but I don't think it would be a hard sell.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/internal/EventImpl.java
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/internal/EventImpl.java:1006]: Do spotless & intellij warn you about null errors if you don't?

Comment (edward-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/internal/EventImpl.java:1006]: You can still use `EventImpl` as a type, sans interface.   Since attributes are not first class citizens, I don't think that `EventImpl` inherits the attributes of the interface through the `@Override` attribute.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:229]: The hip was approved.  This, I assume, would require another HIP at this point.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:320]: adding in next commit.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/internal/EventImpl.java
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/internal/EventImpl.java:1006]: 1. https://stackoverflow.com/questions/62435938/should-i-mark-both-the-interface-and-its-implementation-with-the-nonnull-annota
2. https://github.com/projectlombok/lombok/issues/807#issuecomment-538710386

**Review by cody-littley — APPROVED** [review]

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java
Comment (hendrikebbers) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:216]: null annotation and check for `in`

**Review by hendrikebbers — APPROVED** [review]

**Review by timo0 — APPROVED** [review]

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java]: fixed in next commit.

**Review by cody-littley — APPROVED** [review]

**Review by lpetrovic05 — APPROVED** [review]

**Review by poulok — APPROVED** [review]

**Pull request merged**
